support-lib/cpp/SharedFuture.hpp

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright 2021 Snap, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+ #pragma once
+
+#include "Future.hpp"
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <variant>
+#include <vector>
+
+namespace djinni {
+
+// SharedFuture is a wrapper around djinni::Future to allow multiple consumers (i.e. like std::shared_future)
+// The API is designed to be similar to djinni::Future.
+template<typename T>
+class SharedFuture {
+public:
+    // Create SharedFuture from Future. Runtime error if the future is already consumed.
+    explicit SharedFuture(Future<T>&& future);
+
+    // Transform into Future<T>.
+    Future<T> toFuture() const {
+        if (await_ready()) {
+            co_return await_resume(); // return stored value directly
+        }
+        co_return co_await SharedFuture(*this); // retain copy during coroutine suspension
+    }
+
+    void wait() const {
+        return [this]() -> Future<void> { co_await *this; }().wait();
+    }
+
+    decltype(auto) get() const {
+        wait();
+        return await_resume();
+    }
+
+    // Transform the result of this future into a new future. The behavior is same as Future::then except that
+    // it doesn't consume the future, and can be called multiple times.
+    template<typename Func>
+    SharedFuture<std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<Func, T>>>> then(Func transform) const {
+        co_return transform(co_await SharedFuture(*this)); // retain copy during coroutine suspension
+    }
+
+    // Overload for T = void or `transform` takes no arugment.
+    template<typename Func, typename = std::enable_if_t<!std::is_invocable_v<Func, T>>>
+    SharedFuture<std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<Func>>>> then(Func transform) const {
```

Comment:
I'm worried it might be surprising to the caller that the functor isn't invoked at all if there is an exception. I think it's a subtle behavior difference that can be easily missed or forgotten:
```c++
sharedFuture.then([] { superImportantCall(); }); // wrong, won't happen if exception!
sharedFuture.then([](auto) { superImportantCall(); }); // correct, will always happen
```
 The alternative isn't much more to write, so I don't think the advantage of adding this overload outweighs the potential confusion.

My preference would be to remove this overload or give it a different name that clearly shows this behavior.

Reply:
Yeah that's fair. Originally, the lambdas took resolved values instead of the futures, so the overload was kind of needed for SharedFuture<void>. Removed for now.